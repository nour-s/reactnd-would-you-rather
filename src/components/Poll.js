import React, { Component } from 'react'

class Poll extends Component {
  
  render(){
    const poll = this.props.poll || {};
    
    return (
      <div>
      <h1>
      {JSON.stringify(poll)}
      </h1>
            </div>
           );
  }
}

export default Poll;