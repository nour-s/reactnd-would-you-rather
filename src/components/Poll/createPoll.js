import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { addNewPoll } from "../../actions/shared";
import "./createPoll.scss";

class CreatePoll extends Component {
	state = {
		poll: { optionOneText: "", optionTwoText: "" }
	};

	handleChangeOptionOne(e) {
		this.setState({
			poll: { ...this.state.poll, optionOneText: e.target.value }
		});
	}

	handleChangeOptionTwo(e) {
		this.setState({
			poll: { ...this.state.poll, optionTwoText: e.target.value }
		});
	}

	handleSubmit(e) {
		e.preventDefault();
		this.props.addNewPoll(this.state.poll);
		this.props.history.push(`/`);
	}

	render() {
		let { poll } = this.state;
		return (
			<div className="create-poll">
				<h1 className="create-poll_title">New Question</h1>
				<h3 className="create-poll_subtitle">Would you rather:</h3>
				<form onSubmit={this.handleSubmit.bind(this)}>
					<input
						id="optionOne"
						placeholder="Enter your option one text"
						value={poll.optionOneText}
						onChange={this.handleChangeOptionOne.bind(this)}
					/>
					<hr className="hr-text" data-content="OR" />
					<input
						id="optionTwo"
						placeholder="Enter your option two text"
						value={poll.optionTwoText}
						onChange={this.handleChangeOptionTwo.bind(this)}
					/>
					<br />
					<button className="create-poll_submit">Submit</button>
				</form>
			</div>
		);
	}
}

function mapDispatchToProps(dispatch) {
	return bindActionCreators({ addNewPoll }, dispatch);
}

export default connect(
	null,
	mapDispatchToProps
)(CreatePoll);
