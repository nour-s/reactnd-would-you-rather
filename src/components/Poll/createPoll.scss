.create-poll {
	min-height: 100px;
	border: 1px solid #006495;
	border-radius: 8px;
	max-width: 400px;
	margin: 30px auto;
	padding: 25px 60px 35px 60px;
	position: relative;

	&:after {
		content: "";
		display: block;
		clear: both;
		height: 50px;
	}

	.create-poll_title {
		margin: 0;
		color: #006495;
		margin-bottom: 20px;
	}

	.create-poll_subtitle {
		margin: 15px 0;
	}

	input {
		&::placeholder {
			color: #3169857e;
		}

		padding: 8px;
		border-radius: 5px;
		border: 1px solid #006495;
		margin: 0;
		width: calc(100% - 20px);
		font-family: unset;
	}

	.create-poll_submit {
		background-color: #fff;
		border: 1px solid #006495;
		border-radius: 5px;
		color: #006495;
		display: block;
		margin: 10px;
		font-weight: 500;
		font-family: unset;
		cursor: pointer;
		width: 70px;
		padding: 7px;
		position: absolute;
		right: 0;
		bottom: 0;

		&:hover {
			color: #fff;
			background-color: #006495;
		}
	}
}

.hr-text {
	line-height: 1em;
	position: relative;
	outline: 0;
	border: 0;
	color: black;
	text-align: center;
	height: 1.5em;
	opacity: 1;
	margin: 15px 0;

	&:before {
		content: "";
		// use the linear-gradient for the fading effect
		// use a solid background color for a solid bar
		background: linear-gradient(
			to right,
			transparent,
			#818078,
			transparent
		);
		position: absolute;
		left: 0;
		top: 50%;
		width: 100%;
		height: 1px;
	}
	&:after {
		content: attr(data-content);
		position: relative;
		display: inline-block;
		color: black;
		padding: 0 0.5em;
		line-height: 1.5em;
		// this is really the only tricky part, you need to specify the background color of the container element...
		color: #818078;
		background-color: #fcfcfa;
	}
}
