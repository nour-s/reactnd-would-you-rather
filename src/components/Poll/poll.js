import React, { Component, Fragment } from "react";
import "./poll.scss";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { voteForOption } from "../../actions/shared";
import UserSummary from "../userSummary";
import PropTypes from "prop-types";

export const PollViewMode = {
	Preview: 0,
	Normal: 1
};

export class Poll extends Component {
	constructor(props) {
		super(props);
		const { poll, authedUser } = props;
		this.state = {
			disabled: false,
			selectedAnswer: poll.optionOne.votes.includes(authedUser)
				? poll.optionOne
				: poll.optionTwo.votes.includes(authedUser)
				? poll.optionTwo
				: undefined
		};
	}

	handleAnswerClick = answer => {
		const { poll } = this.props;
		answer = ["optionOne", "optionTwo"][answer - 1];
		this.props.voteForOption({
			pollId: poll.id,
			answer
		});
		this.setState({ disabled: true, selectedAnswer: answer });
		this.props.onPollAnswered();
	};

	answerComponent = (allVotes, vote, isSelectedAnswer) => {
		return (
			<div className={`answer ${isSelectedAnswer ? "selected" : ""}`}>
				<div className="answer_text">{vote.text}</div>
				<div className="answer_details">
					{vote.votes.length} Votes{" "}
					{((vote.votes.length * 100) / allVotes.length).toFixed(0)} %
				</div>
			</div>
		);
	};

	optionComponent = (option, answerClicked) => {
		return (
			<button
				disabled={this.state.disabled && "disabled"}
				onClick={answerClicked}
				className="answer"
			>
				{option.text}
			</button>
		);
	};

	handlePollClick = e => {
		this.props.history.push(`/questions/${this.props.poll.id}`);
	};

	render() {
		const { poll, viewMode } = this.props;
		const allVotes = [...poll.optionOne.votes, ...poll.optionTwo.votes];
		const { selectedAnswer } = this.state;

		return (
			<div
				className={[
					"poll",
					viewMode === PollViewMode.Preview ? "poll--preview" : "",
					this.state.disabled ? "answered" : ""
				].join(" ")}
				onClick={
					viewMode === PollViewMode.Preview
						? e => this.handlePollClick(e)
						: null
				}
			>
				<h2>Would you rather</h2>
				<div className="poll_author">
					<UserSummary user={poll.user} />
				</div>
				<div className="poll_details">
					{!selectedAnswer && (
						<Fragment>
							{this.optionComponent(poll.optionOne, () =>
								this.handleAnswerClick(1)
							)}
							{this.optionComponent(poll.optionTwo, () =>
								this.handleAnswerClick(2)
							)}
						</Fragment>
					)}
					{selectedAnswer && (
						<Fragment>
							{this.answerComponent(
								allVotes,
								poll.optionOne,
								poll.optionOne === selectedAnswer
							)}
							{this.answerComponent(
								allVotes,
								poll.optionTwo,
								poll.optionTwo === selectedAnswer
							)}
						</Fragment>
					)}
				</div>
			</div>
		);
	}
}

Poll.propTypes = {
	poll: PropTypes.object.isRequired,
	authedUser: PropTypes.shape({
		id: PropTypes.string
	}).isRequired,
	voteForOption: PropTypes.func.isRequired,
	onPollAnswered: PropTypes.func.isRequired,
	history: PropTypes.object.isRequired,
	viewMode: PropTypes.object.isRequired
};

function mapDispatchToProps(dispatch) {
	return bindActionCreators({ voteForOption }, dispatch);
}

const mapStateToProps = state => {
	return {
		authedUser: state.authedUser.id
	};
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(Poll);
