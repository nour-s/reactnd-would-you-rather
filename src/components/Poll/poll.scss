.poll {
	min-height: 100px;
	border: 1px solid #006495;
	border-radius: 8px;
	margin: 10px auto;
	padding: 9px;
	display: flex;
	flex-wrap: wrap;

	&:before {
		content: "";
		right: 100%;
	}

	&.answered {
		transform: translateZ(0);
		backface-visibility: hidden;
		position: relative;
		overflow: hidden;
		&:before {
			content: "";
			position: absolute;
			right: 0;
			z-index: -1;
			left: 0;
			bottom: 0;
			background: #2980b9;
			height: 4px;
			transition-duration: 1.7s;
			transition-property: right;
			transition-timing-function: ease-out;
		}
	}

	&.poll--preview {
		&:hover {
			cursor: pointer;
			border-width: 2px;
		}
	}

	h2 {
		margin: 0 0 5px 10px;
		flex-basis: 100%;
	}

	.poll_author {
		flex-basis: 22%;
		display: inline-block;
	}

	.poll_details {
		flex-basis: 60%;
		margin: 0 auto;
	}

	.poll_view-link {
		text-decoration: none;
		padding: 3px;
		text-align: center;
		width: 50%;
		display: block;
		color: #334b53;
		border-radius: 5px;
		margin: 0 auto;
		border: 1px solid #222;
		&:hover {
			border-color: #fff;
			background-color: #334b53;
			color: #fff;
		}
	}

	.answer {
		width: 100%;
		padding: 5px;
		border: 1px solid #006495;
		margin: 0 7px 10px 0;
		background-color: #fff;
		border-radius: 5px;
		cursor: pointer;
		font-size: 14px;
		min-height: 35%;

		.answer_text {
			padding: 5px;
			border-bottom: 1px solid #eee;
		}

		.answer_details {
			padding: 5px;
		}

		&.selected {
			border: 1px solid #f00;
		}
	}

	button.answer {
		&:hover {
			background-color: #006495;
			color: #fff;
		}
	}
}
