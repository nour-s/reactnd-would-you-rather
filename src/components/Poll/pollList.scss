.pollList {
	width: 400px;
	margin: 0 auto;

	.tab-selector {
		width: 100%;

		button {
			text-transform: none;
			width: calc(50% - 10px - 2px - 3px);
			display: inline-block;
			text-decoration: none;
			color: #000;
			border: 1px solid #eee;
			padding: 5px;
			text-align: center;
			border-radius: 3px;
			margin-right: 3px;
			background-color: #fff;
			cursor: pointer;

			&.selected {
				border: 2px solid #eb6;
			}
		}
	}
}
