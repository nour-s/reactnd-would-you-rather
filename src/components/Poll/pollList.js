import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Poll from "./poll";
import "./pollList.scss";
import { switchTab } from "../../actions/shared";

class PollList extends Component {
	handleTabClick = tab => {
		this.props.switchTab(tab);
	};

	getFilteredPolls = tab => {
		const userId = this.props.authedUser.id;
		const polls = Object.keys(this.props.polls).map(
			k => this.props.polls[k]
		);
		switch (tab) {
			case "unanswered":
				return polls.filter(
					p =>
						p.optionOne.votes.indexOf(userId) < 0 &&
						p.optionTwo.votes.indexOf(userId) < 0
				);
			case "answered":
				return polls.filter(
					p =>
						p.optionOne.votes.indexOf(userId) >= 0 ||
						p.optionTwo.votes.indexOf(userId) >= 0
				);
			default:
				return [];
		}
	};

	render() {
		const { ui } = this.props;
		console.log("ui.selectedTab ", ui.selectedTab);
		const polls = this.getFilteredPolls(ui.selectedTab) || [];
		return (
			<div className="pollList">
				<div className="tab-selector">
					<a
						className={
							ui.selectedTab === "unanswered" ? "selected" : ""
						}
						onClick={() => this.handleTabClick("unanswered")}
					>
						Unanswered
					</a>
					<a
						className={
							ui.selectedTab === "answered" ? "selected" : ""
						}
						onClick={() => this.handleTabClick("answered")}
					>
						Answered
					</a>
				</div>
				{polls.map(p => (
					<Poll key={p.id} poll={p} />
				))}
			</div>
		);
	}
}

function mapStateToProps({ authedUser, polls, ui }) {
	return { authedUser, polls, ui };
}

function mapDispatchToProps(dispatch) {
	return bindActionCreators(
		{
			switchTab
		},
		dispatch
	);
}

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(PollList);
