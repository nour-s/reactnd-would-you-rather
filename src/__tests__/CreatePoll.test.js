import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import { CreatePoll } from "../components/Poll/createPoll";

describe("CreatePoll", () => {
	const history = { push: () => {} };
	const addNewPoll = () => {};

	it("Render CreatePoll without crashing", () => {
		const renderer = new ShallowRenderer();
		renderer.render(
			<CreatePoll history={history} addNewPoll={addNewPoll} />
		);
		const tree = renderer.getRenderOutput();
		expect(tree).toMatchSnapshot();
	});
});
