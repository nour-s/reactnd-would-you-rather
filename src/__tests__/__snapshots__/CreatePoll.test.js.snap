// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePoll Render CreatePoll without crashing 1`] = `
<div
  className="create-poll"
>
  <h1
    className="create-poll_title"
  >
    New Question
  </h1>
  <h3
    className="create-poll_subtitle"
  >
    Would you rather:
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      id="optionOne"
      onChange={[Function]}
      placeholder="Enter your option one text"
      value=""
    />
    <hr
      className="hr-text"
      data-content="OR"
    />
    <input
      id="optionTwo"
      onChange={[Function]}
      placeholder="Enter your option two text"
      value=""
    />
    <br />
    <button
      className="create-poll_submit"
    >
      Submit
    </button>
  </form>
</div>
`;
